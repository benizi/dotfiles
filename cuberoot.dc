# precision = 50
50k

# output = hex
16o

## for debugging
# p = print stack with newlines around it
[10an f 10an]sp

# g = debuG
[]sg
# [lpx]sg

## helpers
# d = duplicate the 'n'th item
[d1+0r-SfRdLfR]sd

# m = product of moduli when divided by primes <= 19
# N = input
# P = product of remainders
[
KSK0k dSN 1SP
0 19 17 13 11 7 5 3 2
[lNr%dSWLP*SPd0<X]sX
lXx
+ LNd-+ LP LKk
]sm

# n = next prime
[1+
# L = Large primes (test the product of mods)
[lmx0=n]sL
# S = Small primes (TODO: this is dumb - find a better pattern)
[d1=n d4=n d6=n d8=n d9=n d10=n d12=n d14=n d15=n d16=n d18=n]sS
d19>S d19<L
]sn

# c = cuberoot (initial guess = n/3)
[d3/llx]sc

# i = implementation of one step
[KSK50k d_3Rd3^3ldx-3 3ldx2^*/- LKk]si

# l = loop
[lix d4ldx-d*v1 2 32^/ lgx <l]sl

# t = truncate
[KSK0k d1 0k/LKk-]st

# T = Thirty-two bit-ify
[
KSK0k
d1/ LKk - KSK0k
2 32^*
1/
LKk
]sT

# h = hash values (first 32 bits of fractional parts of the square roots of
# first 8 primes
[lnx
# d [p'=]n n 10an
d v
# d [h'=]n n 10an
[h]n lHn [ := 0x]n
lTx lZx n 10an
lH1+dsH 8>h]sh

# k = round constants
[
lnx
[k]n OSO10o lIn LOo [ := 0x]n
lcx
# d
lTx lZx n 10an
# OSO10o n 10an LOo
lI1+dsI 64>k]sk

# Z = leading zeros
[
d SA
ISIAi16i 10000000 LIi SB
[0n lA lB 16/ dsB >P]sP
lA lB >P
]sZ

# M = main
[
# print hash values
c 1 lhx
# print round constants
c 1 lkx
]sM

lMx

# cuberoot(n) =
#  x' = x-(x^3-n)/(3*x^2)
#  x' = x - a / b
#  x' = x x 3 ^ n - 3 x 2 ^ * / -

# start = n x
# d d 3 ^ = n x x (x^3)
#   4R x x (x^3) n
#   d  x x (x^3) n n
#  _5R n x x (x^3) n
# 4Rd_5R = n x x (x^3) n
# - 3 = n x x ((x^3) - n) 3
# 3Rd_4R = n x x ((x^3) - n) 3 x
# 2 ^ * / -

# [d_3Rd3^3Rd_4R-3 3Rd_4R2^*/-]sa

# x x               3*2^
# x b
# x b n x             3^
# x b n (x^3)          -
# x b a                /
# x (a/b)              -
# x'
# x x 2 ^ 3 * n x 3 ^ - / -
# n x n x x 2 ^ 3 * 3 ^ - / -
# x b n x 3 ^ - / -

# starting with n x
# d2^3* = n x b
# 3Rd_4R 3Rd_4R = n x b n x
# 3^ = n x b n (x^3)
# r- = n x b a
# r/- = n x'
