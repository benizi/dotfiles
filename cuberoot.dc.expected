h0 := 0x6A09E667
h1 := 0xBB67AE85
h2 := 0x3C6EF372
h3 := 0xA54FF53A
h4 := 0x510E527F
h5 := 0x9B05688C
h6 := 0x1F83D9AB
h7 := 0x5BE0CD19
k0 := 0x428A2F98
k1 := 0x71374491
k2 := 0xB5C0FBCF
k3 := 0xE9B5DBA5
k4 := 0x3956C25B
k5 := 0x59F111F1
k6 := 0x923F82A4
k7 := 0xAB1C5ED5
k8 := 0xD807AA98
k9 := 0x12835B01
k10 := 0x243185BE
k11 := 0x550C7DC3
k12 := 0x72BE5D74
k13 := 0x80DEB1FE
k14 := 0x9BDC06A7
k15 := 0xC19BF174
k16 := 0xE49B69C1
k17 := 0xEFBE4786
k18 := 0x0FC19DC6
k19 := 0x240CA1CC
k20 := 0x2DE92C6F
k21 := 0x4A7484AA
k22 := 0x5CB0A9DC
k23 := 0x76F988DA
k24 := 0x983E5152
k25 := 0xA831C66D
k26 := 0xB00327C8
k27 := 0xBF597FC7
k28 := 0xC6E00BF3
k29 := 0xD5A79147
k30 := 0x06CA6351
k31 := 0x14292967
k32 := 0x27B70A85
k33 := 0x2E1B2138
k34 := 0x4D2C6DFC
k35 := 0x53380D13
k36 := 0x650A7354
k37 := 0x766A0ABB
k38 := 0x81C2C92E
k39 := 0x92722C85
k40 := 0xA2BFE8A1
k41 := 0xA81A664B
k42 := 0xC24B8B70
k43 := 0xC76C51A3
k44 := 0xD192E819
k45 := 0xD6990624
k46 := 0xF40E3585
k47 := 0x106AA070
k48 := 0x19A4C116
k49 := 0x1E376C08
k50 := 0x2748774C
k51 := 0x34B0BCB5
k52 := 0x391C0CB3
k53 := 0x4ED8AA4A
k54 := 0x5B9CCA4F
k55 := 0x682E6FF3
k56 := 0x748F82EE
k57 := 0x78A5636F
k58 := 0x84C87814
k59 := 0x8CC70208
k60 := 0x90BEFFFA
k61 := 0xA4506CEB
k62 := 0xBEF9A3F7
k63 := 0xC67178F2
